### DISCUSSION TITLE
How to secure id from browser editing?

### DISCUSSION THREAD

**Discussion by dcbndev** [livewire/livewire#4164] [1. Help] [answered]
Hello,
just started to study livewire, seen Caleb's screencast and few other courses on youtube. I've run into a question, which for me is a huge security hole, and Im asking for help, as I don't know how to solve this.
Let's say we have a posts table. We create a component with a table that displays all posts and add an action button. When that action button is clicked, a modal with the post details is shown. This modal also has a button that performs an action, for example, it updates post title to a given string.
`    public function details($id)
    {

            $this->selectedPost = Post::find($id);
            $this->dispatchBrowserEvent('opendetailsmodal');
    }

    public function change()
    {
        $data = [
            'title' => 'custom title',
        ];
        Post::find($this->selectedPost->id)->update($data);

    }`
The modal is triggered with wire:click, with id passing into livewire component.
`<button wire:click="details({{ $post->id }})">`
In the modal, I retrieve data with  {{ $selectedPost->title ?? '' }} and have a change button for this data, which triggers change() function and changes the title of the post that is loaded in the modal.
`<button wire:click="change()" type="button"`

QUESTION: How to secure the id of the details, as if I change the id in inspector (wire:click="details(1)") and click on it, it will load the data of indicated id, and if I click again the button with wire:click="change()", it will update the model with given id, which is a huge security risk. How to secure the id in this case?

Thanks

**Comment by joshhanley** [comment, marked as answer]
@dcbndev you should secure the ID in the same way you would with a standard Laravel application or API.

You will need to run authorisation on the model, to ensure the user has permissions to view that model, before assigning it to your public property.

The best way to do that is with Laravel's policies.

You can then use Laravel's policies in a Livewire component by including the `AuthorizesRequests` trait https://laravel-livewire.com/docs/2.x/authorization

And then making use of the controller helpers https://laravel.com/docs/8.x/authorization#via-controller-helpers

So your method would look something like this

```php
public function details($id)
{
        $post = Post::find($id);
        
        $this->authorize('view', $post); // Authorise here that the user can view the post before assiging to `$this->selectedPost`
        
        $this->selectedPost = $post;
        
        $this->dispatchBrowserEvent('opendetailsmodal');
}
```

I'd also not recommend calling `Post::find` again in your change method, as you already have the post assigned to `$this->selectedPost` which would have been found in your method above. Also add another authorisation check here to ensure the user can update the post

```php
public function change()
{
    $this->authorize('update', $this->selectedPost);
    
    $data = [
        'title' => 'custom title',
    ];
    
    $this->selectedPost->update($data);
}
```

Hope this helps!

**Reply by dcbndev** [reply]
Thank you @joshhanley for this amazing explanation, it really helps newbies like me. After implementing and testing, I saw that now I can secure posts, filtering by author id. May I have another question? The problem with change remains in case the post author is the same user as logged in user. I mean, if I change the id of the button with detail and then click on it, it will open the modal with the post id that has been modified in inspector. Is there a way to filter this case?

**Reply by joshhanley** [reply]
@dcbndev no worries!

Sure!

Yeah would also use the policies for that, to ensure the logged in user is the author and as such has permission to submit the change.

So in your policy, you would have something like this

```php
public function update(User $user, Post $post)
{
    return $user->id === $post->user_id;
}
```

That way the authorisation check in the change method will fail if the user has attempted to modify a post they don't have permission to.

Hope this helps!
